format.h

```diff
@@ -2628,16 +2639,39 @@ class BasicArrayWriter : public BasicWriter<Char> {
 typedef BasicArrayWriter<char> ArrayWriter;
 typedef BasicArrayWriter<wchar_t> WArrayWriter;
 
+template<class Elem, class Traits = std::char_traits<Elem> >
+class basic_formatbuf : public std::basic_streambuf<Elem, Traits> {
+
+  typedef typename std::basic_streambuf<Elem, Traits>::int_type int_type;
+
+  Buffer<Elem>& buffer_;
+
+public:
+    basic_formatbuf(BasicFormatter<Elem> &formatter) : buffer_(formatter.writer().buffer()) {
+      setp(buffer_.ptr_, buffer_.ptr_ + buffer_.size_, buffer_.ptr_ + buffer_.capacity_);
+  }
+
+  virtual int_type overflow(int_type _Meta = Traits::eof()) {
+    buffer_.grow(buffer_.capacity_ * 2);
+    setp(buffer_.ptr_, buffer_.ptr_ + buffer_.size_, buffer_.ptr_ + buffer_.capacity_);
+
+    return traits_type::to_int_type(*pptr());
+  }
+
+  int_type flush() {
+    buffer_.size_ = pptr() - pbase();
+    return traits_type::to_int_type(*pptr());
+  }
+};
+
 // Formats a value.
 template <typename Char, typename T>
-void format(BasicFormatter<Char> &f, const Char *&format_str, const T &value) {
-  std::basic_ostringstream<Char> os;
-  os << value;
-  std::basic_string<Char> str = os.str();
-  internal::Arg arg = internal::MakeValue<Char>(str);
-  arg.type = static_cast<internal::Arg::Type>(
-        internal::MakeValue<Char>::type(str));
-  format_str = f.format(format_str, arg);
```

Comment:
While I haven't looked in depth at all the errors that are occurring, based on some of the error messages, it seems likely that it's because of the removal of this line. It seems like fixing this may require using a temporary internal::MemoryBuffer that's used by basic_formatbuf, and then format a BasicStringRef<Char> that's extracted from the MemoryBuffer. Since MemoryBuffer can avoid memory allocations for short strings, it seems like this would meet part of the goals for this issue, even if there would still be some unnecessary copies in the "empty format string" case.